ocrd_network/ocrd_network/processing_server.py

```diff
@@ -670,3 +687,41 @@ async def list_processors(self) -> List[str]:
             unique_only=True
         )
         return processor_names_list
+
+    # TODO: think about providing arguments in another way
+    # TODO: this function "just" writes to the queue and returns. A network-client functionality
+    #       should be available to react to everys processors callback. With this feedback
+    #       a blocking mechanism could be provided to inform about starting the cain and waiting for
+    #       the processors to finish and printing when reponses are received from the processors
+    async def run_workflow(self, workflow: UploadFile, workspace_path, callback_url=None) -> List:
+        # core cannot create workspaces by api, but processing-server needs the workspace in the
+        # database. Here the workspace is created if the path available and not existing in db:
+        #from pudb import set_trace; set_trace()
+        if not await create_db_workspace(workspace_path):
+            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                                detail=f"Workspace with path: '{workspace_path}' not existing")
+
+        workflow = (await workflow.read()).decode("utf-8")
+        try:
+            tasks_list = workflow.splitlines()
+            tasks = [ProcessorTask.parse(task_str) for task_str in tasks_list if task_str.strip()]
+        except BaseException as e:
+            print(e)
+            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                                detail=f"Error parsing tasks: {e}")
+        outputs = []
+        last_job_id = ""
+        for task in tasks:
+            data = PYJobInput(
+                agent_type='worker',
+                path_to_mets=workspace_path,
+                input_file_grps=task.input_file_grps,
+                output_file_grps=task.output_file_grps,
+                parameters=task.parameters,
+                callback_url=callback_url,
+                depends_on=[last_job_id] if last_job_id else [],
+            )
+            output = await self.push_processor_job(task.executable, data)
+            outputs.append(output)
+            last_job_id = output.job_id
```

Comment:
Also, the `page_id` attribute is missing completely right now. Is there a reason not to do page-wise processing?

Pseudo-algorithm:

```python
page_range = expand_page_ids(request.page_id) if request.page_id else workspace.physical_pages
depends_on = None
for task in tasks:
  for page in page_range:
                data = PYJobInput(
                agent_type='worker',
                path_to_mets=workspace_path,
                input_file_grps=task.input_file_grps,
                output_file_grps=task.output_file_grps,
                parameters=task.parameters,
                page_id=page
                callback_url=callback_url,
                depends_on=[depends_on] if depends_on else [],
            )
```

As I said, pseudo-code, just to illustrate the idea.

Reply:
Doing page-wise processing and just processing a sub-domain of pages is different, right? It feels like when the `page_id` is provided in the `run_workflow` method, only the range of `page_id` must be processed. This should not affect whether it is page-wise or not. I have implemented that now.

For page-wise processing, it would be better to introduce a flag that enables/disables page-wise processing.  If set, for each page a separate processing request would be created (as in your example), if not set, a single request with the `page_id` domain.

Reply:
> We should put that into a separate function, that generates a list of dicts, so we can test the behavior properly.

A list of `PYJobInput` or `PYJobOutput`? If the former, then it is a bit tricky since the `job_id` is not produced yet by the Processing Server. Hence, we cannot do dependence chaining based on `job_id`. If the latter, then it is doable.

Reply:
> For page-wise processing, it would be better to introduce a flag that enables/disables page-wise processing. If set, for each page a separate processing request would be created (as in your example), if not set, a single request with the page_id domain.

We should then also consider what happens if no `page_id` is provided, but the flag for page-wise processing is set. How do we get the total amount of pages of a workspace? Simply using `workspace.physical_pages` would be inefficient since we are interested only in the number of pages, not the specific page object.

Reply:
This is now implemented but with a few changes:
- for page-wise processing, it is a must for the user to specify the `page_id` when the `page_wise` flag is set. Utilizing `workspace.physical_pages` could have unpredictable side effects when used together with the mets server.
- the outer loop is based on the `page` not on the `task` since the other way around messes up the dependency relation among jobs. It creates a dependency both in pages and tasks, but there is no tight dependency among the pages since we already have an output file group locking mechanism for that.